src/dune_shared_cache/dune_shared_cache.ml

```diff
@@ -250,15 +249,16 @@ struct
             in
             missing, (target, error) :: errors
         in
-        Path.Build.Map.foldi
-          (Targets.Produced.all_files produced_targets)
-          ~init:([], [])
-          ~f:(fun target () -> process_target target)
+        let missing, errors =
+          Nonempty_list.to_list errors |> List.fold_left ~init:([], []) ~f:process_target
+        in
+        List.rev missing, List.rev errors
       in
       (match missing, errors with
        | [], [] ->
          Code_error.raise
-           "compute_target_digests_or_raise_error: spurious target digest failure"
+           "compute_target_digests_or_raise_error: this is impossible because we should \
+            at least be showing the original error"
```

Comment:
I'm surprised we haven't managed to get rid of this unreachable code.

Reply:
We probably could if we had a function like:

```
Nonempty_list.partition_map
:   'a Nonempty_list.t
-> f:('a -> ('b, 'c) Either.t) 
-> ('a Nonempty_list.t * 'b list, 'a list * 'b Nonempty_list.t) Either.t
```

But it seems overkill

Reply:
I did simplify this code further with `List.partition_map` and added a comment as to why this case is hard to eliminate statically https://github.com/ocaml/dune/pull/8731